### PR TITLE
Tighten arm32/arm64 write barrier kill reg sets

### PR DESCRIPTION
**Pull request by BruceForstall**
Also, clean up arm32 NYI_ARM: remove or add asserts for
those that are not possible to encounter, add comments
for some that might be implemented later.

### CODE REVIEW

**Comment by BruceForstall** [comment]
@dotnet-bot test Windows_NT arm64 Cross Checked jitstress1 Build and Test
@dotnet-bot test Windows_NT arm64 Cross Checked jitstress2 Build and Test
@dotnet-bot test Windows_NT arm64 Cross Checked normal Build and Test
@dotnet-bot test Windows_NT arm64 Cross Debug normal Build
@dotnet-bot test Windows_NT arm64 Cross Release normal Build and Test
@dotnet-bot test Windows_NT arm Cross Checked jitstress1 Build and Test
@dotnet-bot test Windows_NT arm Cross Checked jitstress2 Build and Test
@dotnet-bot test Windows_NT arm Cross Checked normal Build and Test
@dotnet-bot test Windows_NT arm Cross Debug normal Build
@dotnet-bot test Windows_NT arm Cross Release normal Build and Test
@dotnet-bot test Ubuntu arm64 Cross Debug normal Build
@dotnet-bot test Ubuntu arm64 Cross Release normal Build and Test

**Comment by BruceForstall** [comment]
@dotnet-bot test Windows_NT arm Cross Checked gcstress0xc Build and Test
@dotnet-bot test Windows_NT arm Cross Checked gcstress0xc_jitstress1 Build and Test
@dotnet-bot test Windows_NT arm Cross Checked gcstress0xc_jitstress2 Build and Test
@dotnet-bot test Windows_NT arm64 Cross Checked gcstress0xc Build and Test
@dotnet-bot test Windows_NT arm64 Cross Checked gcstress0xc_jitstress1 Build and Test
@dotnet-bot test Windows_NT arm64 Cross Checked gcstress0xc_jitstress2 Build and Test

File: src/jit/target.h [hunk 1]
```diff
@@ -1674,26 +1669,65 @@ typedef unsigned short regPairNoSmall; // arm: need 12 bits
...
+  // CORINFO_HELP_ASSIGN_BYREF (JIT_ByRefWriteBarrier):
+  //     On entry:
+  //       x13: the source address (points to object reference to write)
+  //       x14: the destination address (object reference written here)
+  //     On exit:
+  //       x12: trashed
+  //       x13: incremented by 8
+  //       x14: incremented by 8
+  //       x15: trashed
+  //       x17: trashed (ip1) if FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP (currently non-Windows)
```
[the remark this answers is not in the document]
Comment (BruceForstall) [reply src/jit/target.h]: I could just always assume x17 is trashed for arm64.

[the remark this answers is not in the document]
Comment (jkotas) [reply src/jit/target.h]: yes, that would be best.

**Comment by BruceForstall** [comment]
@dotnet-bot test Windows_NT arm64 Cross Checked jitstress1 Build and Test
@dotnet-bot test Windows_NT arm64 Cross Checked jitstress2 Build and Test
@dotnet-bot test Windows_NT arm64 Cross Checked normal Build and Test
@dotnet-bot test Windows_NT arm64 Cross Debug normal Build
@dotnet-bot test Windows_NT arm64 Cross Release normal Build and Test
@dotnet-bot test Windows_NT arm Cross Checked jitstress1 Build and Test
@dotnet-bot test Windows_NT arm Cross Checked jitstress2 Build and Test
@dotnet-bot test Windows_NT arm Cross Checked normal Build and Test
@dotnet-bot test Windows_NT arm Cross Debug normal Build
@dotnet-bot test Windows_NT arm Cross Release normal Build and Test
@dotnet-bot test Ubuntu arm64 Cross Debug normal Build
@dotnet-bot test Ubuntu arm64 Cross Release normal Build and Test
@dotnet-bot test Windows_NT arm Cross Checked gcstress0xc Build and Test
@dotnet-bot test Windows_NT arm Cross Checked gcstress0xc_jitstress1 Build and Test
@dotnet-bot test Windows_NT arm Cross Checked gcstress0xc_jitstress2 Build and Test
@dotnet-bot test Windows_NT arm64 Cross Checked gcstress0xc Build and Test
@dotnet-bot test Windows_NT arm64 Cross Checked gcstress0xc_jitstress1 Build and Test
@dotnet-bot test Windows_NT arm64 Cross Checked gcstress0xc_jitstress2 Build and Test

**Comment by BruceForstall** [comment]
There are many, many diffs. On desktop, SPMI arm64 altjit shows about 1% size improvement for affected methods.

**Comment by BruceForstall** [comment]
@dotnet/arm32-contrib @dotnet/arm64-contrib @dotnet/jit-contrib PTAL

File: src/jit/lowerarmarch.cpp [hunk 2]
```diff
@@ -63,107 +63,84 @@ bool Lowering::IsCallTargetInRange(void* addr)
...
-            case GT_LT:
-            case GT_LE:
-            case GT_GE:
-            case GT_GT:
-                if (childNode->IsIntegralConst(0))
-                {
-                    // TODO-ARM-Cleanup: not tested yet.
-                    NYI_ARM("ARM IsContainableImmed for floating point type");
-
-                    return true;
```
Comment (sdmaclea) [inline src/jit/lowerarmarch.cpp]: @BruceForstall Why is this deleted for arm64?  

For arm 64 this could have been `OperIs(GT_EQ, GT_NE, GT_LT, GT_LE, GT_GE, GT_GT)`

Comment (BruceForstall) [reply src/jit/lowerarmarch.cpp]: This is https://github.com/dotnet/coreclr/pull/17083. It doesn't actually get invoked. I changed it to an assert and never saw the assert hit. In a constructed example, the GT_EQ type came in as TYP_INT, not TYP_FLOAT/DOUBLE, so didn't go down this code path.

Comment (sdmaclea) [reply src/jit/lowerarmarch.cpp]: I thought it looked familiar

**Comment by BruceForstall** [comment]
test Windows_NT arm Cross Checked gcstress0xc_jitstress1 Build and Test
test Windows_NT arm Cross Checked gcstress0xc_jitstress2 Build and Test
test Windows_NT arm64 Cross Checked gcstress0xc Build and Test
test Windows_NT arm64 Cross Checked gcstress0xc_jitstress1 Build and Test
test Windows_NT arm64 Cross Checked gcstress0xc_jitstress2 Build and Test

**Comment by BruceForstall** [comment]
test Windows_NT arm Cross Checked gcstress0xc_jitstress1 Build and Test
test Windows_NT arm Cross Checked gcstress0xc_jitstress2 Build and Test
test Windows_NT arm64 Cross Checked gcstress0xc Build and Test
test Windows_NT arm64 Cross Checked gcstress0xc_jitstress1 Build and Test
test Windows_NT arm64 Cross Checked gcstress0xc_jitstress2 Build and Test

**Comment by BruceForstall** [comment]
test Windows_NT arm Cross Checked gcstress0xc Build and Test

**Comment by BruceForstall** [comment]
Replaced by #17372
